runtime/libtapasco/src/allocator.rs

```diff
@@ -0,0 +1,296 @@
+/*
+ * Copyright (c) 2014-2020 Embedded Systems and Applications, TU Darmstadt.
+ *
+ * This file is part of TaPaSCo
+ * (see https://github.com/esa-tu-darmstadt/tapasco).
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU Lesser General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program. If not, see <http://www.gnu.org/licenses/>.
+ */
+
+use crate::device::DeviceAddress;
+use crate::device::DeviceSize;
+use crate::tlkm::tlkm_ioctl_alloc;
+use crate::tlkm::tlkm_ioctl_free;
+use crate::tlkm::tlkm_mm_cmd;
+use core::fmt::Debug;
+use snafu::ResultExt;
+use std::fs::File;
+use std::os::unix::prelude::*;
+use std::sync::Arc;
+
+#[derive(Debug, Snafu, PartialEq)]
+pub enum Error {
+    #[snafu(display("No memory of size {} available.", size))]
+    OutOfMemory { size: DeviceSize },
+    #[snafu(display("Invalid memory size {}.", size))]
+    InvalidSize { size: DeviceSize },
+    #[snafu(display("Invalid memory alignment {}.", alignment))]
+    InvalidAlignment { alignment: DeviceSize },
+    #[snafu(display("Can't free unknown memory address {}.", ptr))]
+    UnknownMemory { ptr: DeviceAddress },
+    #[snafu(display("Could not free memory: {}", source))]
+    IOCTLFree { source: nix::Error },
+}
+type Result<T, E = Error> = std::result::Result<T, E>;
+
+pub trait Allocator: Debug {
+    fn allocate(&mut self, size: DeviceSize) -> Result<DeviceAddress>;
+    fn free(&mut self, ptr: DeviceAddress) -> Result<()>;
+}
+
+#[derive(Debug, Getters, Copy, Clone)]
+struct MemoryFree {
+    base: DeviceAddress,
+    size: DeviceSize,
+}
+
+#[derive(Debug, Getters)]
+pub struct GenericAllocator {
```

Comment:
Some comments would be nice. Especially for the difference between GenericAllocator and DriverAllocator

Reply:
Yes! Comments are missing almost everywhere right now...